### DISCUSSION TITLE
Linking failure in asm.js: Unexpected stdlib member

### DISCUSSION THREAD

**Discussion by batuhansozer** [openpgpjs/openpgpjs#1227] [Q&A] [answered]
(node:8760) V8: /home/zelos/Desktop/Kedi/kedichat/chat/node_modules/openpgp/dist/openpgp.js:2491 Linking failure in asm.js: Unexpected stdlib member

I get this error every time I use openpgp

**Comment by chesnokovilya** [comment]
Hello, could you please include node version, openpgp version, and a way to replicate this bug? It does not happen on other systems.

**Comment by twiss** [comment]
From the error message, it sounds like this could happen if you're extending a built-in object (of the standard library) with a new function or property (member). Are you (or a package you're importing) writing anything to `global`, or perhaps the `prototype` of a built-in function?

In particular, the global object that's passed to the asm function at that line is `Uint8Array`. Could you run
```js
console.log(Uint8Array.toString());
console.log(Object.getOwnPropertyDescriptors(Uint8Array.prototype));
```
and post the output here?

**Comment by twiss** [comment]
Let me know if you still have this issue.

**Comment by dan-turner** [comment]
I'm seeing this too:

```
› node
Welcome to Node.js v12.16.1.
Type ".help" for more information.
> Uint8Array.toString()
'function Uint8Array() { [native code] }'
> Object.getOwnPropertyDescriptors(Uint8Array.prototype)
{
  constructor: {
    value: [Function: Uint8Array],
    writable: true,
    enumerable: false,
    configurable: true
  },
  BYTES_PER_ELEMENT: { value: 1, writable: false, enumerable: false, configurable: false }
}
```

**Comment by twiss** [comment]
Hey @dan-turner :wave: Thanks! That looks perfectly normal.

I'm assuming you're using OpenPGP.js in a larger Node.js project, though? Could you run the same two `console.log` lines just before calling OpenPGP.js? One of your other dependencies might be messing with the built-in `Uint8Array`.

If you still get the same result, could you try to create a minimal reproducible example?

For what it's worth, openpgp works fine for me in that version of Node:

```
$ node
Welcome to Node.js v12.16.1.
Type ".help" for more information.
> const openpgp = require('openpgp')
undefined
> openpgp.encrypt({ message: openpgp.message.fromText('hi'), passwords: 'test' }).then(console.log, console.error)
Promise { <pending> }
> {
  data: '-----BEGIN PGP MESSAGE-----\r\n' +
    'Version: OpenPGP.js v4.10.1\r\n' +
    'Comment: https://openpgpjs.org\r\n' +
    '\r\n' +
    'wy4ECQMIGlkVb2cEWZ3gNg4uusRwkPb26fM5FahTT4HjJMQULb3s7woNnCZg\r\n' +
    'ALXd0joBbjdz0db3CFkiC7f5P5xdf4d9IFrBJy+OMbt9nXZxUFCgdn42T9Rx\r\n' +
    'phomGifIs4kMFmrQBwEaF9bJ\r\n' +
    '=B/0x\r\n' +
    '-----END PGP MESSAGE-----\r\n'
}
```

**Comment by subha-sfcu** [comment]
> From the error message, it sounds like this could happen if you're extending a built-in object (of the standard library) with a new function or property (member). Are you (or a package you're importing) writing anything to `global`, or perhaps the `prototype` of a built-in function?
> 
> In particular, the global object that's passed to the asm function at that line is `Uint8Array`. Could you run
> 
> ```js
> console.log(Uint8Array.toString());
> console.log(Object.getOwnPropertyDescriptors(Uint8Array.prototype));
> ```
> 
> and post the output here?

> console.log(Uint8Array.toString());
function Uint8Array() { [native code] }
undefined
> console.log(Object.getOwnPropertyDescriptors(Uint8Array.prototype));
{
  constructor: {
    value: [Function: Uint8Array],
    writable: true,
    enumerable: false,
    configurable: true
  },
  BYTES_PER_ELEMENT: { value: 1, writable: false, enumerable: false, configurable: false }
}
undefined
>

**Comment by nottoseethesun** [comment]
Getting this exact error (`node_modules/openpgp/dist/openpgp.js:2491 Linking failure in asm.js: Unexpected stdlib member`) as well; test given above works fine as expected (see below).   What are ideas as to how to fix, or is this an actual bug in this project, the NodeJS `openpgp`?

```bash
$ node --version
v14.11.0
```

```typescript
console.log(Uint8Array.toString());
console.log(Object.getOwnPropertyDescriptors(Uint8Array.prototype));
```
yields:

```
   function Uint8Array() { [native code] }

   {
      constructor: {
        value: [Function: Uint8Array],
        writable: true,
        enumerable: false,
        configurable: true
      },
      BYTES_PER_ELEMENT: { value: 1, writable: false, enumerable: false, configurable: false }
    }
```

Also the repl test works fine:

```
$ node
Welcome to Node.js v14.11.0.
Type ".help" for more information.
> const openpgp = require('openpgp');
undefined
> openpgp.encrypt({ message: openpgp.message.fromText('hi'), passwords: 'test' }).then(console.log, console.error);
Promise { <pending> }
> {
  data: '-----BEGIN PGP MESSAGE-----\r\n' +
    'Version: OpenPGP.js v4.10.4\r\n' +
    'Comment: https://openpgpjs.org\r\n' +
    '\r\n' +
    'wy4ECQMIVnCjm+2GQhLg0T2Cun09/wp284sd7bJ206ewRDtec0u0VCI7ayJD\r\n' +
    'xVi20joBRD8rFaQQvAtlJe2jeLnCUPq4AHAoLQV3/SNLzIOQRxtC2s9q9qIQ\r\n' +
    'JT/LCjIC/U/h0KJgxAAJ3YuT\r\n' +
    '=Dmhy\r\n' +
    '-----END PGP MESSAGE-----\r\n'
}


```

**Comment by twiss** [comment, marked as answer]
Hey :wave: If it works in the REPL but not in your application, then my best guess is that there's something about the environment you're running the application in that makes the globals different from the expected ones. E.g. just a wild guess but are you using Jest? The issue could be https://github.com/facebook/jest/issues/2549.

Alternatively, maybe there's another dependency of your application that's messing with the globals, or somehow interacting badly with OpenPGP.js?

**Reply by nottoseethesun** [reply]
Yep using Jest - thanks!

**Reply by twiss** [reply]
Alrighty :+1: In that case, this might help: https://github.com/openpgpjs/openpgpjs/discussions/1226#discussioncomment-354266

**Comment by Janaka-Steph** [comment]
I tried using `https://www.npmjs.com/package/jest-environment-node-single-context` as suggested in facebook/jest#2549 but didn't work for me

**Comment by dan-turner** [comment]
Still occurring.  The tests do pass but still results in the warning.  This is the output when running with the suggest `--trace-warnings` flag:

```
› yarn node --trace-warnings ./node_modules/.bin/jest keyring 
yarn node v1.22.11
(node:43860) V8: /Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:2491 Linking failure in asm.js: Unexpected stdlib member
    at Sha1.Hash.acquire_asm (/Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:2427:76)
    at Sha1.Hash (/Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:2422:14)
    at new Sha1 (/Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:3411:47)
    at Function.Sha1.bytes (/Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:3418:16)
    at SecretKey.PublicKey.getFingerprintBytes (/Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:39017:34)
    at SecretKey.PublicKey.getFingerprint (/Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:39027:48)
    at SecretKey.PublicKey.getKeyId (/Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:39000:59)
    at Signature.sign (/Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:39999:26)
    at /Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:33187:27
    at async Promise.all (index 0)
(node:43860) V8: /Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:1083 Linking failure in asm.js: Unexpected stdlib member
    at wrapper (/Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:1590:6)
    at AES_ECB.AES.acquire_asm (/Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:1679:59)
    at AES_ECB.AES (/Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:1674:14)
    at new AES_ECB (/Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:2079:23)
    at new C (/Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:25529:21)
    at Object.wrap (/Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:25199:15)
    at Object.encrypt (/Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:29539:39)
    at Object.publicKeyEncrypt (/Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:27177:22)
    at PublicKeyEncryptedSessionKey.encrypt (/Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:39205:20)
    at /Users/dan/code/my-api/node_modules/openpgp/dist/openpgp.js:35845:7
```
